File: apps/agents/src/runtime.ts
import { Duration, Effect } from "effect";

import {
  asRecord,
  type DispatchEnvelope,
  type RuntimeOutputLine,
  runtimeContractVersion,
} from "./types";

function toError(error: unknown): Error {
  return error instanceof Error ? error : new Error(String(error));
}

function assertEnvelope(value: unknown): DispatchEnvelope {
  if (!value || typeof value !== "object") {
    throw new Error("dispatch envelope must be an object");
  }

  const envelope = value as Partial<DispatchEnvelope>;
  if (typeof envelope.version !== "string" || envelope.version.length === 0) {
    throw new Error("dispatch envelope version is required");
  }
  if (
    !envelope.run ||
    typeof envelope.run.id !== "string" ||
    typeof envelope.run.project_id !== "string"
  ) {
    throw new Error("dispatch envelope run is invalid");
  }
  if (
    !envelope.agent ||
    typeof envelope.agent.id !== "string" ||
    typeof envelope.agent.slug !== "string" ||
    typeof envelope.agent.model !== "string"
  ) {
    throw new Error("dispatch envelope agent is invalid");
  }
  if (
    !envelope.deployment ||
    typeof envelope.deployment.id !== "string" ||
    typeof envelope.deployment.provider !== "string"
  ) {
    throw new Error("dispatch envelope deployment is invalid");
  }
  if (
    !envelope.callback ||
    typeof envelope.callback.base_url !== "string" ||
    typeof envelope.callback.run_id !== "string" ||
    typeof envelope.callback.run_token !== "string"
  ) {
    throw new Error("dispatch envelope callback is invalid");
  }

  return envelope as DispatchEnvelope;
}

export function parseEnvelope(
  raw: string
): Effect.Effect<DispatchEnvelope, Error> {
  return Effect.try({
    try: () => assertEnvelope(JSON.parse(raw)),
    catch: toError,
  });
}

export function serializeOutputLine(output: RuntimeOutputLine): string {
  return output.kind === "raw" ? output.line : JSON.stringify(output.event);
}

export function buildRuntimeOutput(
  envelope: DispatchEnvelope
): Effect.Effect<readonly RuntimeOutputLine[]> {
  const payload = asRecord(envelope.payload);
  const config = asRecord(envelope.agent.config);

  const scenario = String(payload._scenario ?? config.scenario ?? "success");
  const delayMs = Number(payload._delay_ms ?? config.delay_ms ?? 0);

  return Effect.gen(function* () {
    if (delayMs > 0) {
      yield* Effect.sleep(Duration.millis(delayMs));
    }

    if (scenario === "invalid_json") {
      return [
        {
          kind: "raw",
          line: "{not-json}",
        },
      ] as const;
    }

    const outputs: RuntimeOutputLine[] = [
      {
        kind: "event",
        event: {
          type: "stream",
          stream_id: "default",
          chunk: `agent:${envelope.agent.slug}:thinking `,
        },
      },
      {
        kind: "event",
        event: {
          type: "checkpoint",
          state: {
            phase: "planning",
            agent_id: envelope.agent.id,
            deployment_id: envelope.deployment.id,
          },
        },
      },
    ];

    if (scenario === "duplicate_checkpoint") {
      outputs.push({
        kind: "event",
        event: {
          type: "checkpoint",
          state: {
            phase: "planning",
            duplicate: true,
          },
        },
      });
    }

    outputs.push(
      {
        kind: "event",
        event: {
          type: "tool_call",
          tool_name: "local.echo",
          input: {
            payload,
          },
          output: {
            echoed: payload.prompt ?? payload.input ?? "ok",
          },
          duration_ms: 5,
          status: "completed",
        },
      },
      {
        kind: "event",
        event: {
          type: "usage",
          provider: "local",
          model: envelope.agent.model || "local-agent",
          prompt_tokens: 12,
          completion_tokens: 8,
          total_tokens: 20,
          cost_microusd: 200,
        },
      },
      {
        kind: "event",
        event: {
          type: "stream",
          stream_id: "default",
          chunk: "done",
          done: true,
        },
      }
    );

    if (scenario === "disconnect") {
      return outputs;
    }

    if (scenario === "fail") {
      outputs.push({
        kind: "event",
        event: {
          type: "fail",
          error: String(payload._error ?? "runtime requested failure"),
        },
      });
      return outputs;
    }

    outputs.push({
      kind: "event",
      event: {
        type: "complete",
        result: {
          ok: true,
          contract_version: runtimeContractVersion,
          agent_id: envelope.agent.id,
          run_id: envelope.run.id,
          payload,
        },
      },
    });

    return outputs;
  });
}

File: apps/strait/internal/agents/runtime.go
package agents

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

const runtimeContractVersion = "v1"

type RuntimeDispatchEnvelope struct {
	Version    string                    `json:"version"`
	Run        RuntimeDispatchRun        `json:"run"`
	Agent      RuntimeDispatchAgent      `json:"agent"`
	Deployment RuntimeDispatchDeployment `json:"deployment"`
	Payload    json.RawMessage           `json:"payload,omitempty"`
	Callback   RuntimeDispatchCallback   `json:"callback"`
	Retry      *RuntimeDispatchRetry     `json:"retry,omitempty"`
}

type RuntimeDispatchRun struct {
	ID          string `json:"id"`
	ProjectID   string `json:"project_id"`
	Attempt     int    `json:"attempt"`
	TimeoutSecs int    `json:"timeout_secs"`
}

type RuntimeDispatchAgent struct {
	ID             string            `json:"id"`
	Slug           string            `json:"slug"`
	Model          string            `json:"model"`
	ModelFallbacks []string          `json:"model_fallbacks,omitempty"`
	Config         json.RawMessage   `json:"config,omitempty"`
	ProviderKeys   map[string]string `json:"provider_keys,omitempty"`
}

type RuntimeDispatchDeployment struct {
	ID             string          `json:"id"`
	Version        int             `json:"version"`
	Provider       string          `json:"provider"`
	ConfigSnapshot json.RawMessage `json:"config_snapshot,omitempty"`
	SandboxPolicy  json.RawMessage `json:"sandbox_policy,omitempty"`
}

type RuntimeDispatchCallback struct {
	BaseURL  string `json:"base_url"`
	RunID    string `json:"run_id"`
	RunToken string `json:"run_token"`
}

type RuntimeDispatchRetry struct {
	LastCheckpoint json.RawMessage `json:"last_checkpoint,omitempty"`
	CheckpointAt   string          `json:"checkpoint_at,omitempty"`
	PreviousError  string          `json:"previous_error,omitempty"`
}

type RuntimeEventType string

const (
	RuntimeEventCheckpoint RuntimeEventType = "checkpoint"
	RuntimeEventUsage      RuntimeEventType = "usage"
	RuntimeEventToolCall   RuntimeEventType = "tool_call"
	RuntimeEventStream     RuntimeEventType = "stream"
	RuntimeEventComplete   RuntimeEventType = "complete"
	RuntimeEventFail       RuntimeEventType = "fail"
)

type RuntimeEvent struct {
	Type RuntimeEventType `json:"type"`

	State json.RawMessage `json:"state,omitempty"`

	Provider         string `json:"provider,omitempty"`
	Model            string `json:"model,omitempty"`
	PromptTokens     int    `json:"prompt_tokens,omitempty"`
	CompletionTokens int    `json:"completion_tokens,omitempty"`
	TotalTokens      int    `json:"total_tokens,omitempty"`
	CostMicrousd     int64  `json:"cost_microusd,omitempty"`

	ToolName   string          `json:"tool_name,omitempty"`
	Input      json.RawMessage `json:"input,omitempty"`
	Output     json.RawMessage `json:"output,omitempty"`
	DurationMs int             `json:"duration_ms,omitempty"`
	Status     string          `json:"status,omitempty"`

	Chunk    string `json:"chunk,omitempty"`
	StreamID string `json:"stream_id,omitempty"`
	Done     bool   `json:"done,omitempty"`

	Result json.RawMessage `json:"result,omitempty"`
	Error  string          `json:"error,omitempty"`
}

type runtimeEventState struct {
	terminal       bool
	completedIDs   map[string]struct{}
	terminalEvent  RuntimeEventType
	terminalResult bool
}

func (s *runtimeEventState) Validate(event *RuntimeEvent) error {
	if event == nil {
		return errors.New("runtime event is required")
	}
	if event.Type == "" {
		return errors.New("runtime event type is required")
	}
	if s.terminal {
		return fmt.Errorf("runtime event %q arrived after terminal event %q", event.Type, s.terminalEvent)
	}
	if s.completedIDs == nil {
		s.completedIDs = make(map[string]struct{})
	}

	switch event.Type {
	case RuntimeEventCheckpoint:
		if len(event.State) == 0 {
			return errors.New("checkpoint event requires state")
		}
	case RuntimeEventUsage:
		if strings.TrimSpace(event.Provider) == "" {
			return errors.New("usage event requires provider")
		}
		if strings.TrimSpace(event.Model) == "" {
			return errors.New("usage event requires model")
		}
	case RuntimeEventToolCall:
		if strings.TrimSpace(event.ToolName) == "" {
			return errors.New("tool_call event requires tool_name")
		}
		if strings.TrimSpace(event.Status) == "" {
			event.Status = "completed"
		}
	case RuntimeEventStream:
		if strings.TrimSpace(event.StreamID) == "" {
			event.StreamID = "default"
		}
		if _, exists := s.completedIDs[event.StreamID]; exists {
			return fmt.Errorf("stream %q received chunks after done", event.StreamID)
		}
		if event.Chunk == "" && !event.Done {
			return errors.New("stream event requires chunk or done=true")
		}
		if event.Done {
			s.completedIDs[event.StreamID] = struct{}{}
		}
	case RuntimeEventComplete:
		s.terminal = true
		s.terminalEvent = event.Type
		s.terminalResult = true
	case RuntimeEventFail:
		if strings.TrimSpace(event.Error) == "" {
			return errors.New("fail event requires error")
		}
		s.terminal = true
		s.terminalEvent = event.Type
		s.terminalResult = true
	default:
		return fmt.Errorf("unknown runtime event type %q", event.Type)
	}

	return nil
}
